storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+			// client does not support this.
+			attachment, err := s.openstackClient.AttachVolume(
+				string(a.Attachment.InstanceId),
+				createdVol.VolumeId,
+				"", // TODO(katco): Fill this in when attachment support is added to parameters.
+			)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			volAtt = append(volAtt, attachment)
+
+		} else {
+
+			createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag.String())
+			if err != nil {
+				return nil, nil, err
+			}
+
+			vols = append(vols, createdVol)
+		}
+	}
+
+	return vols, volAtt, nil
+}
+
+// DescribeVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DescribeVolumes(volIds []string) ([]storage.Volume, error) {
+
+	// In most cases, it is quicker to get all volumes and loop
+	// locally than to make several round-trips to the provider.
+	cinderVols, err := s.openstackClient.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	// Map volume ID to volume info for faster lookup.
+	idToCinderVol := make(map[string]storage.Volume)
+	for _, volInfo := range cinderVols {
+		idToCinderVol[volInfo.VolumeId] = volInfo
+	}
+
+	volumes := make([]storage.Volume, len(volIds))
+	for volIdx, volId := range volIds {
+		cinderVol, ok := idToCinderVol[volId]
+		if !ok {
+			return nil, errors.Errorf("could not find volume: %s", volId)
+		}
+		volumes[volIdx] = cinderVol
+	}
+
+	return volumes, nil
+}
+
+// DestroyVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
+			errors[volIdx] = err
+		}
+	}
+	return errors
+}
+
+// ValidateVolumeParams implements storage.VolumeSource.
+func (s *openstackVolumeSource) ValidateVolumeParams(params storage.VolumeParams) error {
+	panic("not implemented")
+}
+
+// AttachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) AttachVolumes(args []storage.VolumeAttachmentParams) ([]storage.VolumeAttachment, error) {
+
+	volAttachments := make([]storage.VolumeAttachment, len(args))
+	for argIdx, attachArg := range args {
+		if attachArg.Provider != OpenstackProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				attachArg.Provider,
+			)
+			continue
+		}
+
+		attachment, err := s.openstackClient.AttachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			"", // TODO(katco): Fill this in when attachment support is added to parameters.
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		volAttachments[argIdx] = attachment
+	}
+
+	return volAttachments, nil
+}
+
+// DetachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DetachVolumes(params []storage.VolumeAttachmentParams) error {
+	panic("not implemented")
+}
+
+type OpenstackAdapter interface {
```

Comment:
Does this need to be exported? Maybe it does, so needs a comment, as do the methods.